android/app/src/main/java/fi/fmi/mobileweather/LargeWidgetProvider.java

```diff
@@ -0,0 +1,173 @@
+
+        // Show normal view
+        main.setInt(R.id.normalLayout, "setVisibility", VISIBLE);
+        // Hide error view
+        main.setInt(R.id.errorLayout, "setVisibility", GONE);
+
+        main.setOnClickPendingIntent(R.id.mainLinearLayout, pendingIntent);
+
+        json = useNewOrStoredJsonObject(json, pref);
+        if (json == null) return;
+
+        if (background.equals("dark")) {
+            setColors(main,
+                    Color.parseColor("#191B22"),
+                    Color.rgb(255, 255, 255));
+        }
+        else if (background.equals("light")) {
+            setColors(main,
+                    Color.rgb(255, 255, 255),
+                    Color.rgb(48, 49, 147));
+        } else {
+            setColors(main,
+                    Color.TRANSPARENT,
+                    Color.rgb(48, 49, 147));
+        }
+        Log.d("Download json", "Forecast json: " + json);
+        Log.d("Download json", "Crisis json: " + json2);
+
+
+        try {
+            // Get the keys of the JSONObject
+            Iterator<String> keys = json.keys();
+
+            // Retrieve the first key
+            if (!keys.hasNext()) {
+                return;
+            }
+            String firstKey = keys.next();
+            Log.d("Download json", "First key (geoid): " + firstKey);
+
+            // Extract the JSONArray associated with the first key
+            JSONArray data = json.getJSONArray(firstKey);
+
+            // get the first 5 JsonObject from the JSONArray
+            for (int i = 0; i < 5; i++) {
+                JSONObject forecast = data.getJSONObject(i);
+
+                if (i == 0) {
+                    // set the location name and region
+                    String name = forecast.getString("name");
+                    String region = forecast.getString("region");
+                    main.setTextViewText(R.id.locationNameTextView, name + ", ");
+                    main.setTextViewText(R.id.locationRegionTextView, region);
+                }
+
+
+                long epochTime = forecast.getLong("epochtime");
+                String temperature = forecast.getString("temperature");
+                String weathersymbol = forecast.getString("smartSymbol");
+
+                // get timeTextView0 or timeTextView1 etc. based on i from main
+                int timeTextViewId = context.getResources().getIdentifier("timeTextView" + i, "id", context.getPackageName());
+                int temperatureTextViewId = context.getResources().getIdentifier("temperatureTextView" + i, "id", context.getPackageName());
+                int weatherIconImageViewId = context.getResources().getIdentifier("weatherIconImageView" + i, "id", context.getPackageName());
+
+                // ** set the time, temperature and weather icon
+
+                LocalDateTime localDateTime = LocalDateTime.ofInstant(Instant.ofEpochSecond(epochTime), ZoneId.systemDefault());
```

Comment:
Can't use device local time for forecast, because should use timezone of the selected location. Meaning if user is has selected New York we use America/New_York timezone and if Tokyo then we use Asia/Tokyo timezone.

You can get localtime from Timeseries API with `localtime` parameter and use it. Or you can use timezone returned from geolocation query.